### PR TITLE
G-API: Introduce ONNX backend for Inference

### PR DESCRIPTION
**Pull request by dmatveev**
- Basic operations are implemented (Infer, -ROI, -List, -List2);
- Implemented automatic preprocessing for ONNX models;
- Test suite is extended with `OPENCV_GAPI_ONNX_MODEL_PATH` env for test data
  (test data is an ONNX Model Zoo repo snapshot);
- Fixed kernel lookup logic in core G-API:
  - Lookup NN kernels not in the default package, but in the associated
    backend's aux package. Now two NN backends can work in the same graph.
- Added Infer SSD demo and a combined ONNX/IE demo;

<cut/>

### How to build:

Build & install the ONNX RT (currently tested with v1.5.1):
```bash
$ git clone --recursive https://github.com/microsoft/onnxruntime.git
$ cd onnxruntime
$ git checkout v1.5.1
$ git submodule update --init
$ ./build.sh --config Release --build_shared_lib --parallel \
$     --cmake_extra_defines CMAKE_INSTALL_PREFIX=install
$ cd build/Linux/Release
$ make install
```

...then specify extra options to OpenCV CMake:
```bash
$ cmake /path/to/opencv -DWITH_ONNX=ON -DORT_INSTALL_DIR=path-to-onnxrt-install
```

### How to test:

Clone ONNX Model Zoo & selected models
```bash
git clone --recursive https://github.com/onnx/models.git
cd models
git lfs pull --include=vision/object_detection_segmentation/ssd-mobilenetv1/model/ssd_mobilenet_v1_10.onnx --exclude=""
git lfs pull --include=vision/classification/squeezenet/model/squeezenet1.0-9.onnx --exclude=""
git lfs pull --include=vision/body_analysis/emotion_ferplus/model/emotion-ferplus-8.onnx --exclude=""
```
When running G-API tests, specify `OPENCV_GAPI_ONNX_MODEL_PATH` environment variable to the model repo root. Expected result:
```
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from ONNX
[ RUN      ] ONNX.Infer
[       OK ] ONNX.Infer (37 ms)
[ RUN      ] ONNX.InferROI
[       OK ] ONNX.InferROI (26 ms)
[ RUN      ] ONNX.InferROIList
[       OK ] ONNX.InferROIList (44 ms)
[ RUN      ] ONNX.Infer2ROIList
[       OK ] ONNX.Infer2ROIList (43 ms)
[----------] 4 tests from ONNX (150 ms total)
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-1
build_image:Custom=ubuntu-onnx:20.04
test_modules:Custom=gapi
test_filter:Custom=*ONNX*
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -775,6 +778,11 @@ if(WITH_QUIRC)
   add_subdirectory(3rdparty/quirc)
   set(HAVE_QUIRC TRUE)
 endif()
+
+if(WITH_ONNX)
+  include(cmake/FindONNX.cmake)
```
Comment (alalek) [inline CMakeLists.txt:783]: Where is this file?

File: cmake/FindONNX.cmake [hunk 2]
```diff
@@ -0,0 +1,31 @@
+ocv_clear_vars(HAVE_ONNX)
+
+set(ORT_INSTALL_DIR "" CACHE PATH "ONNX Runtime install directory")
```
Comment (alalek) [inline cmake/FindONNX.cmake:3]: `INSTALL_DIR` is usually used to specify location where to install something.

Consider renaming to `ONNX_ROOT_DIR` (do not use `ONNX_DIR` / `ONNX_ROOT` as they are reserved by CMake).

Comment (alalek) [inline cmake/FindONNX.cmake:3]: We should try `ONNXConfig.cmake` from onnxruntime first through `find_package(ONNX)`.

File: cmake/FindONNX.cmake [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  find_path(ORT_INCLUDE onnxruntime_cxx_api.h
+    ${ORT_INSTALL_DIR}/include/onnxruntime/core/session
+    CMAKE_FIND_ROOT_PATH_BOTH)
+endif()
+
+if(ORT_LIB AND ORT_INCLUDE)
+  set(HAVE_ONNX TRUE)
+  # For CMake output only
+  set(ONNX_LIBRARIES "${ORT_LIB}" CACHE STRING "ONNX Runtime libraries")
+  set(ONNX_INCLUDE_DIR "${ORT_INCLUDE}" CACHE STRING "ONNX Runtime include path")
```
Comment (alalek) [inline cmake/FindONNX.cmake:18]:
```.diff
-ONNX_INCLUDE_DIR
+ONNX_INCLUDE_DIRS
```

---

Also remove CACHE from both vars (not interned to be specified by user due to precondition above).

File: CMakeLists.txt (hunk 1 above)
Comment (dmatveev) [reply CMakeLists.txt:783]: Missed :( Added

File: cmake/FindONNX.cmake (hunk 2 above)
Comment (dmatveev) [reply cmake/FindONNX.cmake:3]: Should I change it right now or after having a custom builder working?

File: cmake/FindONNX.cmake [hunk 4]
```diff
@@ -0,0 +1,31 @@
+ocv_clear_vars(HAVE_ONNX)
+
+set(ORT_INSTALL_DIR "" CACHE PATH "ONNX Runtime install directory")
+
+if(ORT_INSTALL_DIR)
+  find_library(ORT_LIB onnxruntime
+    ${ORT_INSTALL_DIR}/lib
+    CMAKE_FIND_ROOT_PATH_BOTH)
```
Comment (alalek) [inline cmake/FindONNX.cmake:8]: > CMAKE_FIND_ROOT_PATH_BOTH

Do we really need that?

File: cmake/FindONNX.cmake [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+endif()
+
+if(ORT_LIB AND ORT_INCLUDE)
+  set(HAVE_ONNX TRUE)
+  # For CMake output only
+  set(ONNX_LIBRARIES "${ORT_LIB}" CACHE STRING "ONNX Runtime libraries")
+  set(ONNX_INCLUDE_DIR "${ORT_INCLUDE}" CACHE STRING "ONNX Runtime include path")
+
+  # Link target with associated interface headers
+  set(ONNX_LIBRARY "ort" CACHE STRING "ONNX Link Target")
```
Comment (alalek) [inline cmake/FindONNX.cmake:21]: > ort

Please don't obfuscate names.

File: cmake/FindONNX.cmake [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  ocv_add_library(${ONNX_LIBRARY} SHARED IMPORTED)
+  set_target_properties(${ONNX_LIBRARY} PROPERTIES
+                        INTERFACE_INCLUDE_DIRECTORIES ${ORT_INCLUDE}
+                        IMPORTED_LOCATION ${ORT_LIB}
+                        IMPORTED_IMPLIB ${ORT_LIB})
```
Comment (alalek) [inline cmake/FindONNX.cmake:26]: after that

```
set(ONNX_LIBRARIES "${ONNX_LIBRARY}")
```

should be applied.

File: modules/gapi/CMakeLists.txt [hunk 7]
```diff
@@ -205,10 +208,20 @@ if(HAVE_PLAIDML)
...
   # Required for htonl/ntohl on Windows
   ocv_target_link_libraries(${the_module} PRIVATE wsock32 ws2_32)
 endif()
 
+if(HAVE_ONNX)
+  ocv_target_link_libraries(${the_module} PRIVATE ${ONNX_LIBRARY})
+  ocv_target_compile_definitions(${the_module} PRIVATE HAVE_ONNX=1)
+  if(TARGET opencv_test_gapi)
+    ocv_target_compile_definitions(opencv_test_gapi PRIVATE HAVE_ONNX=1)
+    ocv_target_link_libraries(opencv_test_gapi PRIVATE ${ONNX_LIBRARY})
```
Comment (alalek) [inline modules/gapi/CMakeLists.txt:222]:
```.diff
-ONNX_LIBRARY
+ONNX_LIBRARIES
```

should be used.

File: cmake/FindONNX.cmake (hunk 2 above)
Comment (alalek) [reply cmake/FindONNX.cmake:3]: Added extra variable into custom builder:
- `-DONNXRUNTIME_ROOT_DIR=/opt/onnxrt`

**Review by alalek — COMMENTED** [review]
There is still confusion with ONNX / ONNXRUNTIME naming:
- WITH_ONNX
- HAVE_ONNX
- cmake/FindONNX.cmake (not the same as `find_package(ONNX)`)

File: cmake/FindONNX.cmake (hunk 4 above)
Comment (dmatveev) [reply cmake/FindONNX.cmake:8]: Have no idea, tbh. Probably it works even without it.

File: cmake/FindONNX.cmake (hunk 5 above)
Comment (dmatveev) [reply cmake/FindONNX.cmake:21]: ok

File: cmake/FindONNX.cmake (hunk 6 above)
Comment (dmatveev) [reply cmake/FindONNX.cmake:26]: Then CMake output will text about our internal name of this library, but won't show its full actual path

File: cmake/FindONNX.cmake (hunk 3 above)
Comment (dmatveev) [reply cmake/FindONNX.cmake:18]: if I remove those from CACHE, would the "3rd party" thing display them?

**Comment by dmatveev** [comment]
@alalek updated

**Review by alalek — APPROVED** [review]
CMake part looks good to me :+1:

**Comment by dmatveev** [comment]
Thanks!

**Pull request merged**